### PR TITLE
[CARBONDATA-2674][Streaming]Streaming with merge index enabled does not consider the merge index file while pruning.

### PR DESCRIPTION
**Pull request by mohammadshahidkhan**
…

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [X] Any interfaces changed?
 None
 - [X] Any backward compatibility impacted?
 None
 - [X] Document update required?
None
 - [X] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       Added test to varify the merge index scenarios.
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
  NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6654/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5481/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6658/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5485/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6698/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5525/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5552/

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java [hunk 1]
```diff
@@ -341,21 +343,32 @@ public CarbonTable getOrCreateCarbonTable(Configuration configuration) throws IO
...
-          if (index.exists()) {
-            // data file exists
-            CarbonIndexFileReader indexReader = new CarbonIndexFileReader();
+          SegmentIndexFileStore segmentIndexFileStore = new SegmentIndexFileStore();
+          segmentIndexFileStore.readAllIIndexOfSegment(segmentDir);
+          Map<String, byte[]> carbonIndexMap = segmentIndexFileStore.getCarbonIndexMap();
+          Set<Map.Entry<String, byte[]>> entries = carbonIndexMap.entrySet();
+          CarbonIndexFileReader indexReader = new CarbonIndexFileReader();
+          for (Map.Entry<String, byte[]> entry : entries) {
+            byte[] value = entry.getValue();
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: give a better name, `value` is too generic

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java [hunk 2]
```diff
@@ -341,21 +343,32 @@ public CarbonTable getOrCreateCarbonTable(Configuration configuration) throws IO
...
+              if (index.exists()) {
+                // data file exists
+                indexReader.openThriftReader(indexPath);
+              }
+            }
             try {
               // map block index
-              indexReader.openThriftReader(indexPath);
-              while (indexReader.hasNext()) {
+              while (null != indexReader && indexReader.hasNext()) {
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: `indexReader` is initialized in line 354, can it be null?

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java [hunk 3]
```diff
@@ -341,21 +343,32 @@ public CarbonTable getOrCreateCarbonTable(Configuration configuration) throws IO
...
-          // index file exists
-          if (index.exists()) {
-            // data file exists
-            CarbonIndexFileReader indexReader = new CarbonIndexFileReader();
+          SegmentIndexFileStore segmentIndexFileStore = new SegmentIndexFileStore();
+          segmentIndexFileStore.readAllIIndexOfSegment(segmentDir);
+          Map<String, byte[]> carbonIndexMap = segmentIndexFileStore.getCarbonIndexMap();
+          Set<Map.Entry<String, byte[]>> entries = carbonIndexMap.entrySet();
+          CarbonIndexFileReader indexReader = new CarbonIndexFileReader();
+          for (Map.Entry<String, byte[]> entry : entries) {
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: It seems key is never used, then it is better to use `carbonIndexMap.values`

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java (hunk 2 above)
Comment (mohammadshahidkhan) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: Thanks @jackylk indexReader can not be null. 
The CarbonIndexFileReader.hasNext() can throw NullPointerException if index file does not exists,
so corrected the hasNext API.
          if (index.exists()) {
                // data file exists
                indexReader.openThriftReader(indexPath);
              }

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java (hunk 3 above)
Comment (mohammadshahidkhan) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: Fixed

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java (hunk 1 above)
Comment (mohammadshahidkhan) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonTableInputFormat.java]: Fixed renamed to fileData

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5550/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5569/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5552/

**Comment by jackylk** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5571/

**Pull request closed without merging**